### PR TITLE
ci: enable unit and functional tests

### PR DESCRIPTION
**Pull request by brendanashworth**
This commit re-enables both the unit tests and functional tests for the
CI machine, so it'll run automatically for PRs and commits. This will
ensure commits don't accidentally break tests and have unintended side
effects.

Functional tests are run against the play.minio.io server.

This commit also fixes issues in the tests that would have caused them
to fail. It also allows functional tests to be run on non-mint machines
and the CI after b3b05e0, which introduced a few bugs.

Note: this was partially done before (no functional tests) in https://github.com/minio/minio-js/commit/1698358cc4c384fa1f9e998f8d0ca0d983b73bbd. It was then reverted from Travis in https://github.com/minio/minio-js/commit/fce7f6a84a588dd05df041c9b65b3c3b1730dc2a, in what I think was an unrelated change. Then it fixes a few bugs from b3b05e0.

cc @harshavardhana

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -74,21 +74,6 @@ gulp.task('lint', function() {
...
-  .pipe(gulp.dest('dist/test'))
-  .on('end', function() {
-    gulp.src('dist/test/functional-tests.js')
-    .pipe(mocha({
-      reporter: 'spec',
-      ui: 'bdd',
-    }))
-  })
-})
-
```
Comment (krishnasrinivas) [inline gulpfile.js]: @brendanashworth sometimes we might want to run just the functional tests, so you can keep this code.

Comment (brendanashworth) [inline gulpfile.js]: sure!

**Comment by brendanashworth** [comment]
@krishnasrinivas fixing a quick bug that deals with simultaneous unit tests. will be ready in a few

**Comment by krishnasrinivas** [comment]
@brendanashworth there are some conflits

**Comment by brendanashworth** [comment]
@krishnasrinivas will fix! Don't merge yet, it still needs a bit of fixing. I'll tell you when it's ready 😄

**Comment by krishnasrinivas** [comment]
> @krishnasrinivas will fix! Don't merge yet, it still needs a bit of fixing. I'll tell you when it's ready

You can put `[WIP]` in the PR subject till it is not ready, like this:
`[WIP] ci: enable unit and functional tests`

**Comment by brendanashworth** [comment]
Okay it's looking good 👍 PTAL @krishnasrinivas @poornas

**Comment by brendanashworth** [comment]
Not sure, but this might fix https://github.com/minio/minio-js/issues/602. It doesn't require a `/mint` directory to run the tests anymore, it'll generate its own data if not given a directory.

Also, this was done on the way towards https://github.com/minio/minio-js/issues/568.

**Comment by krishnasrinivas** [comment]
> Not sure, but this might fix #602. It doesn't require a /mint directory to run the tests anymore, it'll generate its own data if not given a directory.

#602 I think needs random-data generating readable streams which does not allocate memory.

**Comment by brendanashworth** [comment]
ping @poornas

**Comment by brendanashworth** [comment]
thanks everyone!

**Pull request merged**